### PR TITLE
feat: protocol rewards

### PR DESCRIPTION
**Pull request by kulkarohan**

### CODE REVIEW

File: foundry.toml [hunk 1]
```diff
@@ -22,20 +22,18 @@ via_ir = true
...
 solc_version = '0.8.17'
 
-
 [rpc_endpoints]
-mainnet = "https://eth-mainnet.g.alchemy.com/v2/${ALCHEMY_KEY}"
 goerli = "https://eth-goerli.g.alchemy.com/v2/${ALCHEMY_KEY}"
+mainnet = "https://eth-mainnet.g.alchemy.com/v2/${ALCHEMY_KEY}"
 # for optimism, since we are just using this for deployment/fork testing,
 # we can use these since they're lower volume.
+base = "https://developer-access-mainnet.base.org"
```
Comment (iainnash) [inline foundry.toml:30]: Should this be a seperate diff?

File: src/nft/ZoraCreator1155Impl.sol [hunk 2]
```diff
@@ -182,7 +200,12 @@ contract ZoraCreator1155Impl is
     /// @param tokenId tokenId to check
     /// @param role role to check for admin
     function _requireAdminOrRole(address user, uint256 tokenId, uint256 role) internal view {
-        if (!(_hasAnyPermission(tokenId, user, PERMISSION_BIT_ADMIN | role) || _hasAnyPermission(CONTRACT_BASE_ID, user, PERMISSION_BIT_ADMIN | role))) {
+        if (
+            !(
+                _hasAnyPermission(tokenId, user, PERMISSION_BIT_ADMIN | role)
```
Comment (iainnash) [inline src/nft/ZoraCreator1155Impl.sol:205]: I think you may need to update your linters – would make a smaller pr as well

**Comment by oveddan** [comment]
could you add some tests that show end to end the effect of create/minter rewards? could you add a description to the PR that explains this change? 

I was looking for that in this PR to see how that works, and it's not clear from the PR description or the comments what a create referral or how mintWithRewards differs.  Tests that show the rewards distribution logic in action would go a long way.

File: src/nft/ZoraCreator1155Impl.sol [hunk 3]
```diff
@@ -248,17 +259,40 @@ contract ZoraCreator1155Impl is
...
-        string memory newURI,
+        string calldata newURI,
         uint256 maxSupply
     ) public onlyAdminOrRole(CONTRACT_BASE_ID, PERMISSION_BIT_MINTER) nonReentrant returns (uint256) {
+        uint256 tokenId = _setupNewTokenAndPermission(newURI, maxSupply, msg.sender, PERMISSION_BIT_ADMIN);
+
+        return tokenId;
+    }
+
+    /// @notice Set up a new token with a create referral
```
Comment (oveddan) [inline src/nft/ZoraCreator1155Impl.sol:270]: could you please explain in this comment what a create referral is and the effect is has?

File: .changeset/little-moose-pay.md [hunk 4]
```diff
@@ -0,0 +1,5 @@
+---
+"@zoralabs/zora-1155-contracts": patch
+---
+
+Add support for rewards
```
Comment (oveddan) [inline .changeset/little-moose-pay.md:5]: this should be a minor version bump since its a new feature

File: src/nft/ZoraCreator1155Impl.sol [hunk 5]
```diff
@@ -2,8 +2,12 @@
 pragma solidity 0.8.17;
 
 import {ERC1155Upgradeable} from "@zoralabs/openzeppelin-contracts-upgradeable/contracts/token/ERC1155/ERC1155Upgradeable.sol";
+import {ReentrancyGuardUpgradeable} from "@zoralabs/openzeppelin-contracts-upgradeable/contracts/security/ReentrancyGuardUpgradeable.sol";
+import {UUPSUpgradeable} from "@zoralabs/openzeppelin-contracts-upgradeable/contracts/proxy/utils/UUPSUpgradeable.sol";
 import {IERC1155MetadataURIUpgradeable} from "@zoralabs/openzeppelin-contracts-upgradeable/contracts/interfaces/IERC1155MetadataURIUpgradeable.sol";
 import {IERC165Upgradeable} from "@zoralabs/openzeppelin-contracts-upgradeable/contracts/interfaces/IERC165Upgradeable.sol";
+import {IProtocolRewards} from "@zoralabs/protocol-rewards/dist/contracts/interfaces/IProtocolRewards.sol";
+import {ERC1155Rewards, ERC1155RewardsStorage} from "@zoralabs/protocol-rewards/dist/contracts/abstract/ERC1155/ERC1155Rewards.sol";
```
Comment (oveddan) [inline src/nft/ZoraCreator1155Impl.sol:10]: this shouldn't come from the dist folder - you should be able to configure your package with `files` including `src` and it would include those files automatically; see:
https://github.com/ourzora/zora-1155-contracts/blob/main/package.json#L29

File: src/nft/ZoraCreator1155Impl.sol [hunk 6]
```diff
@@ -386,6 +420,37 @@ contract ZoraCreator1155Impl is
...
+    /// @param tokenId The token ID to mint
+    /// @param quantity The quantity of tokens to mint
+    /// @param minterArguments The arguments to pass to the minter
+    /// @param mintReferral The referrer of the mint
+    function mintWithRewards(
+        IMinter1155 minter,
+        uint256 tokenId,
+        uint256 quantity,
+        bytes calldata minterArguments,
+        address mintReferral
```
Comment (oveddan) [inline src/nft/ZoraCreator1155Impl.sol:440]: how about just including this in minterArguments? then we wouldn't need the function overload and could save on contract size; in the `mint` function you could check for the presence of `mintReferral` in the `minterArguments`

File: src/nft/ZoraCreator1155Impl.sol [hunk 7]
```diff
@@ -386,6 +420,37 @@ contract ZoraCreator1155Impl is
...
+    ) external payable nonReentrant {
+        // Require admin from the minter to mint
+        _requireAdminOrRole(address(minter), tokenId, PERMISSION_BIT_MINTER);
+
+        // Get value sent and handle mint rewards
+        uint256 ethValueSent = _handleRewardsAndGetValueSent(msg.value, quantity, getCreatorRewardRecipient(), mintReferral, createReferrals[tokenId]);
```
Comment (oveddan) [inline src/nft/ZoraCreator1155Impl.sol:446]: hard to follow what this code does when looking in this PR, given it comes from another repo and there aren't any new tests around this new functionality in this PR. could you add some tests that demonstrate the effect of this new functionality?

**Review by oveddan — CHANGES_REQUESTED** [review]

File: package.json [hunk 8]
```diff
@@ -32,8 +32,9 @@
     "package/"
   ],
   "dependencies": {
-    "@openzeppelin/contracts": "4.8.3",
+    "@openzeppelin/contracts": "^4.9.2",
     "@zoralabs/openzeppelin-contracts-upgradeable": "^4.8.4",
+    "@zoralabs/protocol-rewards": "^1.0.1",
```
Comment (oveddan) [inline package.json:37]: CI is failing cause this would need to be public npm package

File: package.json [hunk 9]
```diff
@@ -32,8 +32,9 @@
     "package/"
   ],
   "dependencies": {
-    "@openzeppelin/contracts": "4.8.3",
+    "@openzeppelin/contracts": "^4.9.2",
```
Comment (oveddan) [inline package.json:35]: this version bump probably belongs in its own PR

**Comment by iainnash** [comment]
i think we just can make it public

File: src/nft/ZoraCreator1155Impl.sol [hunk 10]
```diff
@@ -623,6 +698,25 @@ contract ZoraCreator1155Impl is
         }
     }
 
+    /// @notice Withdraws ETH from the Zora Rewards contract
+    function withdrawRewards(uint256 amount) public onlyAdminOrRole(CONTRACT_BASE_ID, PERMISSION_BIT_FUNDS_MANAGER) {
```
Comment (oveddan) [inline src/nft/ZoraCreator1155Impl.sol:702]: does this need to be only called by an admin or `PERMISSION_BIT_FUNDS_MANAGER` since it always withdraws to the fundsRecipient?

File: src/nft/ZoraCreator1155Impl.sol [hunk 11]
```diff
@@ -623,6 +698,25 @@ contract ZoraCreator1155Impl is
...
+        bytes memory data = abi.encodeWithSelector(IProtocolRewards.withdraw.selector, amount);
```
Comment (oveddan) [inline src/nft/ZoraCreator1155Impl.sol:703]: what exactly is happening here? is the eth transferred first to this contract, then to the funds recipient? can you add a comment explaining this logic?

File: src/nft/ZoraCreator1155Impl.sol (hunk 10 above)
Comment (oveddan) [inline src/nft/ZoraCreator1155Impl.sol:702]: is having an `amount` necessary? how about just withdrawing the full balance? it could make things simpler

File: src/version/ContractVersionBase.sol [hunk 12]
```diff
@@ -8,6 +8,6 @@ import {IVersionedContract} from "../interfaces/IVersionedContract.sol";
 contract ContractVersionBase is IVersionedContract {
     /// @notice The version of the contract
     function contractVersion() external pure override returns (string memory) {
-        return "1.3.3";
+        return "1.3.4";
```
Comment (oveddan) [inline src/version/ContractVersionBase.sol:11]: this version bump should happen via the version packages pr

File: src/nft/ZoraCreator1155Impl.sol [hunk 13]
```diff
@@ -623,6 +698,25 @@ contract ZoraCreator1155Impl is
...
+
+        (bool success, ) = address(PROTOCOL_REWARDS).call(data);
+
+        if (!success) {
+            revert();
+        }
+
+        address fundsRecipient = config.fundsRecipient;
+
+        if (fundsRecipient != address(0)) {
```
Comment (oveddan) [inline src/nft/ZoraCreator1155Impl.sol:713]: this should revert if funds recipient is address(0) - otherwise the funds will get locked in this contract

File: test/nft/ZoraCreator1155.t.sol [hunk 14]
```diff
@@ -584,30 +611,408 @@ contract ZoraCreator1155Test is Test {
...
+        target.addPermission(tokenId, address(simpleMinter), adminRole);
 
         vm.expectRevert(abi.encodeWithSelector(IZoraCreator1155.CannotMintMoreTokens.selector, tokenId, 1001, 0, 1000));
         vm.prank(admin);
-        target.mint(minter, tokenId, 1001, abi.encode(recipient));
+        target.mint(simpleMinter, tokenId, 1001, abi.encode(recipient));
     }
 
-    function test_callSale() external {
+    function test_FreeMintRewards(uint256 quantity) public {
```
Comment (oveddan) [inline test/nft/ZoraCreator1155.t.sol:622]: thanks for adding all these tests!!!

File: test/fee/MintFeeManager.t.sol [hunk 15]
```diff
@@ -14,13 +15,15 @@ import {SimpleMinter} from "../mock/SimpleMinter.sol";
 contract MintFeeManagerTest is Test {
```
Comment (oveddan) [inline test/fee/MintFeeManager.t.sol:15]: is `MintFeeManager even used anymore?  Isn't all the logic now in `ProtocolRewards` and if so can we delete the `MintFeeManager` class?

**Review by jgeary — APPROVED** [review]

**Review by oveddan — APPROVED** [review]

**Review by iainnash — APPROVED** [review]
